cmd/jujud/updateseries/updateseries.go

```diff
@@ -118,38 +127,76 @@ func (c *UpdateSeriesCommand) Run(ctx *cmd.Context) error {
 		return err
 	}
 
+	jujuVersion, err := agent.GetJujuVersion(c.machineAgent, c.dataDir)
+	if err != nil {
+		return err
+	}
+
+	var (
+		startedUnitNames   []string
+		startedMachineName string
+	)
 	switch {
 	case toInitSys == service.InitSystemUpstart:
 		return errors.NewNotSupported(nil, "downgrade to series using upstart not supported")
 	case toInitSys == fromInitSys && toInitSys == service.InitSystemSystemd:
-		if err = c.copyTools(); err != nil {
+		if err := c.manager.CopyAgentBinary(c.machineAgent, c.unitAgents, c.dataDir, c.toSeries, c.fromSeries, jujuVersion); err != nil {
 			return err
 		} else {
-			ctx.Infof("successfully copied tools and relinked agent tools")
+			ctx.Infof("successfully copied agent binaries and relinked binaries")
 		}
 		if !c.startAgents {
 			break
 		}
-		if err = c.startAllAgents(ctx); err != nil {
+
+		if startedMachineName, startedUnitNames, err := c.manager.StartAllAgents(c.machineAgent, c.unitAgents, c.dataDir, c.toSeries); err != nil {
 			return err
+		} else {
+			for _, unitName := range startedUnitNames {
+				ctx.Infof("started %s service", unitName)
+			}
+			ctx.Infof("started %s service", startedMachineName)
 		}
 		ctx.Infof("all agents successfully restarted")
 	case toInitSys == service.InitSystemSystemd:
+		var (
+			startedSysdServiceNames []string
+			startedSymServiceNames  []string
+		)
 		errorHappened := false
-		if err = c.writeSystemdAgents(ctx); err != nil {
+		failedAgentNames = failedAgentNames[:0]
+		if startedSysdServiceNames, startedSymServiceNames, failedAgentNames, err = c.manager.WriteSystemdAgents(c.machineAgent, c.unitAgents, c.dataDir, systemdDir, systemdMultiUserDir, c.toSeries); err != nil {
+			for _, agentName := range failedAgentNames {
+				ctx.Warningf("failed to write service for %s: %s", agentName, err)
+			}
 			ctx.Warningf("%s", err)
 			errorHappened = true
+		} else {
+			for _, sysSvcName := range startedSysdServiceNames {
+				ctx.Infof("wrote %s agent, enabled and linked by systemd", sysSvcName)
+				for _, symSvcName := range startedSymServiceNames {
+					ctx.Infof("wrote %s agent, enabled and linked by symlink", symSvcName)
+				}
+			}
 		}
-		if err = c.copyTools(); err != nil {
+		for _, symSvcName := range startedSymServiceNames {
+			ctx.Infof("wrote %s agent, enabled and linked by symlink", symSvcName)
+		}
+		if err = c.manager.CopyAgentBinary(c.machineAgent, c.unitAgents, c.dataDir, c.toSeries, c.fromSeries, jujuVersion); err != nil {
 			ctx.Warningf("%s", err)
 			errorHappened = true
 		} else {
-			ctx.Infof("successfully copied tools and relinked agent tools")
+			ctx.Infof("successfully copied agent binaries and relinked binaries")
 		}
 		switch {
 		case !errorHappened && c.startAgents:
-			if err = c.startAllAgents(ctx); err != nil {
+			if startedMachineName, startedUnitNames, err = c.manager.StartAllAgents(c.machineAgent, c.unitAgents, c.dataDir, c.toSeries); err != nil {
```

Comment:
Same here - this would be clearer as:
```go
startedMachineName, startedUnitNames, err := c.manager.StartAllAgents(
    c.machineAgent, c.unitAgents, c.dataDir, c.toSeries,
)
if err != nil {
```

Basically if you're assigning several things *and* then checking the error return, it makes the line too wide. It's better to split it up in that case - we try to keep lines under 120 chars. (I'd prefer to keep them under about 80 but I'm a bit weird!)